src/pyhf/modifiers/shapefactor.py

```diff
@@ -62,64 +63,71 @@ def __init__(self, shapefactor_mods, pdfconfig, mega_mods):
         and at that point can be used to compute the effect of shapefactor.
         """
 
+        self.batch_size = batch_size
         pnames = [pname for pname, _ in shapefactor_mods]
         keys = ['{}/{}'.format(mtype, m) for m, mtype in shapefactor_mods]
         shapefactor_mods = [m for m, _ in shapefactor_mods]
 
-        self._parindices = list(range(len(pdfconfig.suggested_init())))
-        self._shapefactor_indices = [
-            self._parindices[pdfconfig.par_slice(p)] for p in pnames
-        ]
+        parfield_shape = (self.batch_size or 1, len(pdfconfig.suggested_init()))
+        self.parameters_helper = ParamViewer(parfield_shape, pdfconfig.par_map, pnames)
+
         self._shapefactor_mask = [
             [[mega_mods[s][m]['data']['mask']] for s in pdfconfig.samples] for m in keys
         ]
+
         global_concatenated_bin_indices = [
-            j for c in pdfconfig.channels for j in range(pdfconfig.channel_nbins[c])
-        ]
-        # compute the max so that we can pad with 0s for the right shape
```

Comment:
these comments got deleted (and they were helpful)... the new change to this code is pretty opaque and I'm having a hard time actually understanding what's going on

Reply:
added extensive comment